Fix TODOs to address memory leaks

gabrielschulhof opened the pull request:
Fixes: https://github.com/nodejs/node-addon-api/issues/333

gabrielschulhof commented:
Finally, I chose to attach the data by creating a property whose key is an anonymous symbol and whose value is an external containing the data. This leaves the `napi_wrap()` slot free, and will not leave any unsightly symbols hanging around on an object once we have ubiquitous support for `napi_add_finalizer()`.

gabrielschulhof commented:
Testing to make sure that the finalizer gets called could be done using assertions made about the output of a coverage tool. Basically, we have to make sure that the [line](https://github.com/nodejs/node-addon-api/pull/343/files#diff-f546e4cc35b454813e5264fe9c9e6fc8R73) where the data gets deleted gets hit a certain number of times in response to the `global.gc()` call [here](https://github.com/nodejs/node-addon-api/pull/343/files#diff-f4e47e66a182f6a2faaf439d468d00f6R16).

I manually verified this by adding an `fprintf(stderr, ...)` to the finalizer and running the `thunking_manual` test via `node --expose-gc test/thunking_manual.js`.

To test this on the CI, we need a code coverage build followed by a code coverage run, followed by parsing of the code coverage output so we can make assertions about when the finalizer gets called.

mhdawson commented on napi.h:
```diff
@@ -1236,66 +1236,86 @@ namespace Napi {
   class PropertyDescriptor {
```
Did not review in depth, but thinking that instead of changing public API we should add new methods, deprecate the old ?  

While the commitment to stability for node-addon-api is lower than N-API itself, at this point I think it would be good to avoid forcing code changes when modules upgrade.

gabrielschulhof commented:
@mhdawson I had forgotten to update the PR to update the documentation. That's done now.

mhdawson commented on test/binding.gyp:
```diff
@@ -39,6 +40,7 @@
       'dependencies': ["<!(node -p \"require('../').gyp\")"],
       'cflags': [ '-Werror', '-Wall', '-Wextra', '-Wpedantic', '-Wunused-parameter' ],
       'cflags_cc': [ '-Werror', '-Wall', '-Wextra', '-Wpedantic', '-Wunused-parameter' ],
+      'defines': ['NODE_ADDON_API_DISABLE_DEPRECATED']
```
Does this mean we won't test the deprecated functions?

gabrielschulhof replied:
Yes, it does mean that.

mhdawson commented:
@gabrielschulhof this is big enough it would probably be good to spend a few minutes together going through it.  Maybe after the N-API meeting on Thursday.

gabrielschulhof commented on napi-inl.h:
```diff
@@ -2777,20 +2822,106 @@ inline T* ObjectWrap<T>::Unwrap(Object wrapper) {
+  // After defining the class we iterate once more over the property descriptors
+  // and attach the data associated with accessors and instance methods to the
+  // newly created JavaScript class.
+  for (size_t idx = 0; idx < props_count; idx++) {
```
@mhdawson I realized that it's OK to always attach this data to the JS class. We do not need to afterwards retrieve the instance method and attach the data to *it*, because we do not need to worry about the scenario where someone "detaches" an instance method from the class' `.prototype` property. The reason for this is that in our implementation of instance methods, we attempt to `napi_unwrap()` the `this` parameter in order to retrieve the instance of the native class, so we can then call the corresponding instance method on the native instance. Since the JavaScript class is garbage-collected, no new JavaScript instances can be created, and so no JavaScript objects can arrive via the `this` property that will successfully `napi_unwrap()` to give us the native instance. This fails nicely with a `napi_object_expected` or a `napi_invalid_arg` status that gets propagated back into JavaScript as an exception.

mhdawson commented on test/object/object.cc:
```diff
@@ -83,20 +92,36 @@ void DefineProperties(const CallbackInfo& info) {
+      PropertyDescriptor::Function(str7, TestFunction),
+#else // NODE_ADDON_API_DISABLE_DEPRECATED
       PropertyDescriptor::Function(env, str7, TestFunction),
+#endif // !NODE_ADDON_API_DISABLE_DEPRECATED
```
I think we want to test both the deprecated and non-deprecated so this should just be #ifndef, #endif as opposed to using #else ?  Same goes to all of the other places as well.

gabrielschulhof commented:
@mhdawson I split the deprecated tests into their own test at `test/object/object_deprecated.{cc,js}` and modified to build to conditionally include `test/object/object_deprecated.cc`. The test runs unconditionally but short-circuits successfully if `binding.object_deprecated` is absent.

mhdawson commented:
Looks good other than the comment about testing should test both deprecated and non-deprecated versions as opposed to one or the other.  Looks like all of the other issues we discussed in person are addressed and I'll sign of land once the test ifdefs are fixed up.

gabrielschulhof commented:
@mhdawson I fixed that. Now, when you run `npm test` it runs both the deprecated and the non-deprecated tests, and when you run `npm test --disable-deprecated` it only runs the non-deprecated tests.

mhdawson commented:
Landed as 917bd60baa2f3ff800b8e63baa9301bec7e65b90
